comp/core/gui/guiimpl/auth.go

```diff
@@ -0,0 +1,135 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2024-present Datadog, Inc.
+
+package guiimpl
+
+import (
+	"crypto/hmac"
+	"crypto/sha256"
+	"encoding/base64"
+	"encoding/binary"
+	"fmt"
+	"strings"
+	"time"
+)
+
+const tokenVersion = "v1"
+
+// authenticator represents an authentication mechanism.
+type authenticator struct {
+	duration   time.Duration // The duration for which the authentication token is valid.
+	signingKey []byte        // The key used for signing the authentication token.
+}
+
+func newAuthenticator(authToken string, duration time.Duration) authenticator {
+	return authenticator{
+		duration:   duration,
+		signingKey: []byte(authToken),
+	}
+}
+
+func (a *authenticator) GenerateAccessToken() string {
+	now := time.Now()
+	return hmacToken(a.signingKey, now, now.Add(a.duration))
+}
+
+func (a *authenticator) ValidateToken(token string) error {
+	// Split the token into the payload and HMAC sum
+	parts := strings.Split(token, ".")
+	if len(parts) != 3 {
+		return fmt.Errorf("invalid token format")
+	}
+
+	// Check token version
+	if parts[0] != tokenVersion {
+		return fmt.Errorf("token version mismatch: got %s, expected %s", parts[0], tokenVersion)
+	}
+
+	// Decode the payload from base64
+	payloadBytes, err := base64.StdEncoding.DecodeString(parts[1])
+	if err != nil {
+		return fmt.Errorf("failed to decode payload: %w", err)
+	}
+
+	// Ensure the payload contains enough bytes for issued and expiration times
+	if len(payloadBytes) < 16 {
+		return fmt.Errorf("invalid payload")
+	}
+
+	// Extract the issued and expiration times from the payload
+	issuedTime := int64(binary.LittleEndian.Uint64(payloadBytes[:8]))
+	expirationTime := int64(binary.LittleEndian.Uint64(payloadBytes[8:]))
+
+	// Decode the HMAC sum from base64
+	hmacSum, err := base64.StdEncoding.DecodeString(parts[2])
+	if err != nil {
+		return fmt.Errorf("failed to decode HMAC sum: %w", err)
+	}
+
+	// Calculate the expected HMAC sum
+	mac := hmac.New(sha256.New, a.signingKey)
+	mac.Write(payloadBytes)
+	expectedHmacSum := mac.Sum(nil)
+
+	// Check if the HMAC sum matches the expected HMAC sum
+	if !hmac.Equal(hmacSum, expectedHmacSum) {
+		return fmt.Errorf("invalid token signature")
+	}
+
+	now := time.Now()
+
+	// Check if the current time is before the issued time
+	if now.Before(time.Unix(issuedTime, 0)) {
+		return fmt.Errorf("token is invalid")
+	}
```

Comment:
Should we move this check before `// Calculate the expected HMAC sum`  that way if the token is expired we do not generate a HMAC sum. I would imagine is the most expensive operation of this function, so if only calculate it if we need it it would be best 😄